tests/pkl_codegen_java/PolymorphismTest.java

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2025 Apple Inc. and the Pkl project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import static org.junit.Assert.*;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.Animals;
+import org.junit.Test;
+import org.pkl.config.java.Config;
+import org.pkl.config.java.ConfigEvaluator;
+import org.pkl.config.java.JavaType;
+import org.pkl.core.ModuleSource;
+
+public class PolymorphismTest {
+
+  @Test
+  public void testPolymorphicDeserialization() throws IOException {
+    // Create a Pkl config with polymorphic animals
+    String pklConfig =
+        """
+        module test
+
+        import "modulepath:/tests/pkl_codegen_java/module/animals.pkl"
+
+        shelter = new animals.Shelter {
+          animals {
+            new animals.Dog {
+              name = "Buddy"
+              breed = "Golden Retriever"
+            }
+            new animals.Cat {
+              name = "Whiskers"
+              color = "Orange"
+            }
+            new animals.Dog {
+              name = "Max"
+              breed = "German Shepherd"
+            }
+          }
+        }
+        """;
```

Comment:
Let's move this out to a `.pkl` file, and read it in as a resource as part of the `java_test` target. Look at how we do this in `tests/pkl_codegen_java/TestPklCodegen.java`.